### ISSUE TITLE
Parallelizing step 1 stage 3 (level 1 ridge)

### ISSUE THREAD

**Issue by bgorm** [rgcgithub/regenie#184]
I have implemented your workflow for parallelizing step 1 (https://github.com/rgcgithub/regenie/wiki/Further-parallelization-for-level-0-models-in-Step-1) and it speeds up step 1 quite a bit. However, there is still a bottleneck at stage 3 when dealing with multiple phenotypes -- it appears to be single-threaded only and processes each phenotype in series. At large N I have found this can take up to an hour per phenotype, which limits the practicality of analyzing dozens of phenotypes simultaneously. Is there any way to process the level 1 regressions per phenotype in parallel? Thanks in advance.

**Comment by freeseek** [comment]
@joellembatchou I will add that I have been able to take the latest regenie binary and implement the two-tiered approach for step 1 and package in a WDL pipeline that can be easily run in Google cloud. I implemented Stage 2 as running across all phenotypes at once on small windows and Stage 3 as running on each phenotype independently. A few comments:

1) I skipped Stage 1 (i.e. I did not use option `--split-l0`) as my pipeline automatically splits the dataset in windows, so no need to have regenie split it for me. This way Stage 2 tasks also were spared from having to localize the whole cohort genotype files

2) I had to rename the output files from Stage 2 as in my implementation they get all called `fit_bin_parallel_job1_l0_Y*` as regenie in each task is unaware of how many windows are present in the genome

3) I had to rename the input files from Stage 3 to `fit_bin_parallel_job*_l0_Y1` as regenie in each task is unaware of how many phenotypes have been tested

Basically I think of the `fit_bin_parallel_job*_l0_Y*` files as elements of a matrix (windows x phenotypes) and Stage 2 generates the rows and Stage 3 consumes the columns.

This approach makes the pipeline amazingly parallel (and cheap, since I can use preemptible computing on Google cloud). However, I feel like the regenie implementation is very NFS-filesystem-oriented and it was quite an effort to get it to work. My personal suggestion is to allow the user to have full control of the input and output filenames rather than include all of the file handling inside the regenie binary, as not all users will run the binaries on a shared filesystem.

**Comment by dvh13** [comment]
@freeseek we are also using a Google Cloud TRE for Genes & Health. similar to Finngen's. uses WDL. do you have your WDL scripts available/on github, would you be happy for us to adapt? (ps I think Finngen also might have done some regenie WDL script work)

**Comment by dvh13** [comment]
found this https://github.com/FINNGEN/regenie-pipelines

**Comment by unepanthere** [comment]
Running step 1 on our nice 104-cores CPU ... with 100 phenotypes ... and BOOM @ Level 1 ridge (stage 3?)
What!?! Only 1 core is used, and that at a rate of ~20 minutes / phenotype... this is going to take 33 hours!

Uh, how do I go about this? run 10 parallel step 1's with 10 phenotypes each?

"Where's the lamb ssssssaaaaaaauuuuuuuuuuuuuuuuuuccccccccceeeeeeeeeeeeeeeeeeee"
  - Gordon Ramsey

**Comment by freeseek** [comment]
You just have to rename the output files from Stage 2 and then you can run each phenotype as a separate job for Stage 3 ... you don't need to split Stage 2 across phenotypes. Then you have to reassemble the output from Stage 3 to make it look to Step 2 as if it was all run as a single job. It is a little tedious but not too difficult
